File: docs/scripts/CellState.js
export default class CellState {
    constructor(state) {
        this.state = state;
        this.observer = null;
    }
    setState(newstate) {
        this.state = newstate;
        if(this.observer) {
            this.observer.execute(newstate);
        }
    }
    setObserver(obj) {
        this.observer = obj;
    }
    getState() {
        return this.state;
    }
    equals(obj) {
        return obj.getState() === this.state;
    }
}

File: src/main/java/model/CellState.java
package model;

import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;

public class CellState {
    private State state;
    private final PropertyChangeSupport squareObservers = new PropertyChangeSupport(this);

    public CellState(State newstate) {
        this.state = newstate;
    }

    public State getState() {
        return state;
    }

    public void setState(State newValue) {
        var oldValue = this.state;
        this.state = newValue;
        this.squareObservers.firePropertyChange("state", oldValue, newValue);
    }

    public void addSquareObservers(PropertyChangeListener pcl) {
        squareObservers.addPropertyChangeListener(pcl);
    }

    @Override
    public String toString() {
        return String.format("%s", this.state);
    }
}
